cityofaustin/vision-zero
# Migrate the VZ User and CR3 API into our VZ ECS cluster with an API gateway setup

## Associated issues

This PR is part of the solution intending to close https://github.com/cityofaustin/atd-data-tech/issues/18154 and it closes https://github.com/cityofaustin/atd-data-tech/issues/18663 as well.

## Moving Parts

This PR moves a fair bit of things around, both in production, staging, and local development. This:

* Deprecate the use of Zappa and λ functions more broadly for the VZ CR3 / User API. 🔥
* Support the use of ECS to host the above API.
* Add various headers to the outgoing responses in pursuit of ISO's request.
* Incorporate the flask app into the local development stack. Start it with `./vision-zero api-up`. It will be available on port `8085` on localhost.
* Add `local-stack-up` and `local-stack-down` to the `vision-zero` orchestration app, to start and stop the 1) DB, 2) `graphql-engine` instance, 3) user / cr3 API. This is intended to be the "normal" command we use to start and stop our local backend stack.
  * Don't forget to update your `zsh` completion functions with the `-s zsh` flag's output -- this makes the orchestration so much easier to use. This applies to `bash` and `tcsh` users as well.
* Add clarity to the environment file of the VZE to allow you to point at local, staging, and production API endpoints, as needed.
* Incorporate the flask app into a docker image, available on docker hub as `atddocker/vz-user-cr3-api`. 
* Include CI to build and restart these services as appropriate.
* Update the flask app to use modern versions of libraries.
* Write out the use of the flask internals to allow for those updates. 


## Testing

* Take a look at the staging and production VZ ECS clusters. 
  * Check out the new, second services running in them. Do you like the names? Anything you'd change?
  * Check out the task definitions for these two new services. Do you like the resource sizing? Naming?

* Spin up your local stack: 
  * Create a `.env` based on `.env.example` using the 1PW entry `Vision Zero zappa_settings` which is a JSON file. Grab the staging values, or reach out to me for a `.env` via `wormhole`. 
  * `./vision-zero local-stack-up` will get your local stack running, which is now 3 docker containers.
  * Fire up your local VZE, and see if it works. Can you list the VZ users?
  * Review the `editor/.env`. Do you see options on how to adjust what flask app your local VZE points at?
  * ‼️ Aim your local VZE at the various endpoints. Everything work? This is testing the actual flask app itself as installed in the new AWS infrastructure. This is the most important thing to test.

* Review the changes to the CI:
  * These are hard to test per se, so give them the eyeball test, and we can confirm they work at merger. 
  * Review the images that are on docker hub; they were built by the CI.
  * Do you approve of the contents of those images? You can install `dive` with homebrew to make it easy to inspect these file manifests.

* ISO's concerns
  * Review my [header choices](https://github.com/cityofaustin/vision-zero/blob/dockerize-api/api/server.py#L73-L103) related to the ISO's requests. Do you agree? Change anything?
  * Consider my wide-open CORS headers. OK with you?

## Documentation
  * This is forth-coming, and will be written in preparation for my skill share next week on the infra side of this work. We can spin it off into an issue if needed, but I want to call it out of scope on this particular issue.

---

#### Ship list
